transaction/std/src/memo/authenticated_sender.rs

```diff
@@ -0,0 +1,150 @@
+    ) -> Self {
+        // The layout of the memo is:
+        // [0-16] address hash
+        // [16-28] unused
+        // [28-44] HMAC
+
+        let mut memo_data = [0u8; 44];
+        memo_data[..16].copy_from_slice(cred.address_hash.as_ref());
+
+        let shared_secret = cred
+            .subaddress_spend_private_key
+            .key_exchange(receiving_subaddress_view_public_key);
+
+        let hmac_value = compute_category1_hmac(
+            shared_secret.as_ref(),
+            tx_out_public_key,
+            Self::MEMO_TYPE_BYTES,
+            &memo_data,
+        );
+        memo_data[28..].copy_from_slice(&hmac_value);
+
+        Self { memo_data }
+    }
+
+    /// Get the sender address hash from the memo
+    pub fn sender_address_hash(&self) -> AddressHash {
+        let bytes: [u8; 16] = self.memo_data[0..16].try_into().unwrap();
+        AddressHash::from(bytes)
+    }
+
+    /// Validate an AuthenticatedSenderMemo
+    ///
+    /// First, the client should look up the sender's Public Address from their
+    /// hash. If it isn't a known contact we won't be able to authenticate
+    /// them.
+    ///
+    /// Then they need to get the view private key corresponding to the
+    /// subaddress that this TxOut was sent to. This is usually our default
+    /// subaddress view private key.
+    ///
+    /// Finally we can validate the memo against these data. The
+    /// tx_out_public_key is also under the mac, which prevents replay
+    /// attacks.
+    ///
+    /// Arguments:
+    /// * sender_address: The public address of the sender. This can be looked
+    ///   up by the AddressHash provided.
+    /// * receiving_subaddress_view_private_key: This is usually our
+    ///   default_subaddress_view_private_key, but should correspond to whatever
+    ///   subaddress recieved this TxOut.
+    /// * tx_out_public_key: The public key of the TxOut to which this memo is
+    ///   attached.
+    ///
+    /// Returns:
+    /// * subtle::Choice(1u8) if validation passed, subtle::Choice(0u8) if hmac
+    ///   comparison failed.
+    ///
+    /// This function is constant-time.
+    pub fn validate(
+        &self,
+        sender_address: &PublicAddress,
+        receiving_subaddress_view_private_key: &RistrettoPrivate,
+        tx_out_public_key: &CompressedRistrettoPublic,
+    ) -> Choice {
+        let mut result = Choice::from(1u8);
+        let sender_address_hash = AddressHash::from(sender_address);
+        result &= sender_address_hash.ct_eq(&self.sender_address_hash());
+
+        let shared_secret =
+            receiving_subaddress_view_private_key.key_exchange(sender_address.spend_public_key());
+
+        let expected_hmac = compute_category1_hmac(
+            shared_secret.as_ref(),
+            tx_out_public_key,
+            Self::MEMO_TYPE_BYTES,
+            &self.memo_data,
+        );
+        let found_hmac: [u8; 16] = self.memo_data[28..].try_into().unwrap();
```

Comment:
I assume there's no benefit to be had from using `copy_from_slice()` instead of `try_into()` in this scenario?

Reply:
the main difference i think is that this way its a one-liner and found_hmac doesn't have to be `mut`, but they would both panic if there is a length mismatch. i can change it to use `copy_from_slice` but right now i don't see a good reason to prefer it?

Reply:
If there's no good reason for it, that's fine---I assume `try_into()` here is actually doing the `copy_from_slice()` anyways, it would just skip the extra branch in `unwrap()` is all.